DirectProgramming/DPC++FPGA/Tutorials/Features/ac_types/ac_int/src/ac_int.cpp

```diff
@@ -0,0 +1,346 @@
+      // template argument since the width is inferred from the width of the
+      // argument x
+      x[0].set_slc(lsb_accessor[0], y[0]);
+
+      // Bits can also be individually set as follows:
+      x[0][3] = 0;
+      x[0][2] = 0;
+      x[0][1] = 0;
+      x[0][0] = 0;
+    });
+  });
+  q.wait();
+}
+
+int main() {
+#if defined(FPGA_EMULATOR)
+  ext::intel::fpga_emulator_selector selector;
+#else
+  ext::intel::fpga_selector selector;
+#endif
+  bool passed = true;
+
+  try {
+    // create the SYCL device queue
+    queue q(selector, dpc_common::exception_handler);
+
+    // Use a fixed initial seed
+    srand(123);
+
+    // Initialize two random ints
+    int t1 = rand();
+    int t2 = rand();
+    // Truncate each of the two ints to 13 bits. The 14th bit is the sign bit.
+    // In this testbench, even though the datatypes are signed, we will be
+    // storing unsigned values to make the testing process simpler.
+    t1 &= (1 << 13) - 1;
+    t2 &= (1 << 13) - 1;
+
+    std::cout << "Arithmetic Operations:\n";
+    // Test adder
+    {
+      // ac_int offers type casting from and to native datatypes
+      ac_int14 a = t1;
+      ac_int14 b = t2;
+      ac_int15 c;
+      TestAdd(q, a, b, c);
+      int c_golden = t1 + t2;
+      // We can check the result of the ac_int addition with the native C int
+      // addition
+      if (c != c_golden) {
+        passed = false;
+        std::cerr << "Addition failed\n";
+      }
+      std::cout << "ac_int: " << a << " + " << b << " = " << c << "\n";
+      std::cout << "int:    " << t1 << " + " << t2 << " = " << c_golden << "\n";
+    }
+
+    // Test multiplier
+    {
+      t1 = rand() & ((1 << 13) - 1);
+      ac_int14 a = t1;
+      ac_int14 b = t2;
+      ac_int28 c;
+      TestMult(q, a, b, c);
+      int c_golden = t1 * t2;
+      // We can check the result of the ac_int multiplication with the native C
+      // int multiplication
+      if (c != c_golden) {
+        passed = false;
+        std::cerr << "Multiplier failed\n";
+      }
+      std::cout << "ac_int: " << a << " * " << b << " = " << c << "\n";
+      std::cout << "int:    " << t1 << " * " << t2 << " = " << c_golden << "\n";
+    }
+
+    // Test divider
+    {
+      t1 = rand() & ((1 << 13) - 1);
+      t2 = rand() % 50;  // Use a small value for the divisor so that the result
+                         // is not 0 or 1
+      ac_int14 a = t1;
+      ac_int14 b = t2;
+      ac_int15 c;
+      TestDiv(q, a, b, c);
+      int c_golden = t1 / t2;
+      // We can check the result of the ac_int division with the native C
+      // int division
+      if (c != c_golden) {
+        passed = false;
+        std::cerr << "divider failed\n";
+      }
+      std::cout << "ac_int: " << a << " / " << b << " = " << c << "\n";
+      std::cout << "int:    " << t1 << " / " << t2 << " = " << c_golden << "\n";
+    }
+
+    std::cout << "\nBitwise Operations:\n";
+    // Shift operator
+    {
+      t1 = rand() & ((1 << 13) - 1);
```

Comment:
does it make sense to put the testbench code into the `TestXXX()` functions? that would make the `main()` function tidier. I believe the main point of separating them in the [hls version](https://gitlab.devtools.intel.com/hldapps/hls/hls-tutorials/-/blob/master/ac_datatypes/ac_int_basic_ops/basic_ops.cpp) was to separate the device code from testbench code, but that abstraction doesn't really work in oneAPI anyway, since the `TestXXX()` functions already contain host code.